File: setcover/solutions/greedy.cpp
#include <iostream>
#include <vector>
#include <set>
#include <sstream>

using namespace std;

struct Set {
    int cost;
    vector<int> elements;
};

set<int> greedySetCover(int n, const vector<Set>& sets) {
    int m = sets.size();

    set<int> uncovered;
    for (int i = 0; i < n; i++) {
        uncovered.insert(i);
    }

    set<int> chosenSets;

    vector<bool> covered(n);

    while (!uncovered.empty()) {
        int bestSet;
        int bestSetCost = 1;
        int newElementsCovered = 0;

        for(int i = 0; i < m; i++) {
            if(chosenSets.find(i) != chosenSets.end()) {
                continue;
            }

            int newElements = 0;
            for(int elem : sets[i].elements) {
                if(!covered[elem]) {
                    newElements++;
                }
            }

            if(sets[i].cost * newElementsCovered < bestSetCost * newElements) {
                newElementsCovered = newElements;
                bestSetCost = sets[i].cost;
                bestSet = i;
            }
        }

        chosenSets.insert(bestSet);

        for(int elem : sets[bestSet].elements) {
            if(!covered[elem]) {
                covered[elem] = true;
                uncovered.erase(elem);
            }
        }
    }

    return chosenSets;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n, m;
    cin >> n >> m;
    string line;
    getline(cin, line);

    vector<Set> sets(m);

    for(int i = 0; i < m; i++) {
        getline(cin, line);
        istringstream ss(line);
        ss >> sets[i].cost;

        int elem;
        while(ss >> elem) {
            sets[i].elements.push_back(elem);
        }
    }

    set<int> chosenSets = greedySetCover(n, sets);
    int totalCost = 0;
    for(auto i : chosenSets) {
        totalCost += sets[i].cost;
    }

    cout << totalCost << '\n';

    for(auto x : chosenSets) {
        cout << x << ' ';
    }

    cout << '\n';
}

File: setcover/solutions/greedy.py
import sys

def read_input(filename):
    with open(filename) as f:
        n, m = map(int, f.readline().split())
        costs = []
        sets = []
        for _ in range(m):
            data = list(map(int, f.readline().split()))
            costs.append(data[0])
            sets.append(set(data[1:]))
    return n, m, costs, sets


def greedy(n, m, sets, costs):
    uncovered = set(range(n))
    chosen_sets = set()
    covered = [False] * n

    total_cost = 0

    while uncovered:
        best_set = -1
        best_set_cost = 1
        new_elements_covered = 0

        for i in range(m):
            if i in chosen_sets:
                continue

            new_elements = 0
            for elem in sets[i]:
                if not covered[elem]:
                    new_elements += 1

            if new_elements > 0:
                if best_set == -1 or costs[i] * new_elements_covered < best_set_cost * new_elements:
                    new_elements_covered = new_elements
                    best_set_cost = costs[i]
                    best_set = i

        if best_set == -1:
            break

        chosen_sets.add(best_set)
        total_cost += costs[best_set]

        for elem in sets[best_set]:
            if not covered[elem]:
                covered[elem] = True
                uncovered.remove(elem)

    return total_cost, chosen_sets


if __name__ == "__main__":
    filename = sys.argv[1]
    n, m, costs, sets = read_input(filename)
    best_cost, best_solution = greedy(n, m, sets, costs)

    print(best_cost)
    print(*best_solution)